apps/labs/posts/automated-accessibility-testing-with-bokeh.md

```diff
@@ -0,0 +1,178 @@
+Accessibility testing in the Bokeh codebase, and the closest we had to that was
+removed sometime in the past because it had been unused and unmaintained for so
+long. The primary goal of my work would thus be to move Bokeh from 0 to 1 and
+give feedback and findings that could be built upon. Essentially, the spark that
+starts the fire.
+
+## How Bokeh works under the hood
+
+Talking about the scope of what I worked on and the accompanying complexities
+wouldn't make as much sense if I didn't explain the mechanism behind Bokeh. To
+the typical user who needs to get work done, you just install the package, type
+some commands as instructed by the docs, and then run. In doing this, if you
+look closely at your file directly you'll realize that a matching HTML file has
+been created and automatically opens in your browser. Voila! You get an
+interactive plot. With a few lines of Python code, you've done some apparent
+magic.
+This is where things start to get fun. At the end of the day, Javascript is what
+gives interactivity to any HTML page, so the fact that you can interact with the
+beautiful data visualization on an HTML page implies the existence of CSS and
+Javascript. This means Python code is responsible for generating HTML, CSS, and
+Javascript. For this to happen with any tool, there needs to be a shared data
+model between what would be the backend (Python) and what would be client-side
+(what you see on the browser). Bokeh has a brilliant way of doing this, and I'll
+attach a picture from their [SciPy 2023
+tutorial](https://www.youtube.com/watch?v=G0Yc3ck4lC8) to illustrate:
+
+<img alt="Image of Bokeh data model flow represented as a flow chart moving from
+Python to JSON, to Javascript, and then to final HTML output"
+src="/posts/automated-accessibility-testing-with-bokeh/bokeh_bokehjs.jpg"
+style={{padding: '10px 20px'}}/>
+
+The implication of this is that virtually all things on the Python side should
+be representable on a Javascript (for the most part).
+
+## Testing tools
+
+For the right blend of power, ease of use, and modernity, my mentor
+[Gabriel](https://github.com/gabalafou) and I decided to use
+[Playwright](https://playwright.dev/) for setting up the testing environment.
+
+We also decided that it might be worthwhile having axe-core in the testing setup
+to further simplify tests that the Axe API could detect, although due to the
+very nature of Bokeh data visualizations, not many things would require
+axe-core. This is because the visualizations end up getting wrapped in an HTML
+canvas with some styling in the [Shadow
+DOM](https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_shadow_DOM),
+which is something axe does not know how to work with. However, certain plots
+(or documents) rely on extra components or widgets like sliders and dropdowns
+which are done outside the canvas and are thus testable with the Axe API.
+
+## Where we currently are
```

Comment:
Let's put all of the headings together.

Automated Accessibility Testing with Bokeh

- What is Bokeh?
- What is accessibility?
- Proposed Work
- How Bokeh works under the hood
- Testing tools
- Where we currently are
- Progressive work going forward
- Credits

The title is pretty good and descriptive. It gives the reader an idea of what this article will be about. The headings taken together should perform a similar function. I think we can make these level 2 headings more descriptive and informative.